src/Sketch.java

```diff
@@ -28,8 +67,73 @@ public void setup() {
      */
     public void draw() {
         background(180, 180, 255);
-        ball1.draw();
-        ball1.move();
+
+        for (Ball ball : ballsArray) {
+            ball.draw();
+            ball.move();
+        }
+
+        // for each of the 4 balls, check if their radius collides with another ball
+        // if it does, bounce the balls away from each other
+        for (int i = 0; i < ballsArray.length; i++) {
+            for (int j = i + 1; j < ballsArray.length; j++) {
+                if (ballsArray[i].collidesWith(ballsArray[j])) {
+                    ballsArray[i].bounce(ballsArray[j]);
+
+                    ballsArray[i].randomizeColors();
+                    ballsArray[j].randomizeColors();
+                }
+            }
+        }
+
+        for (int i = 0; i < ballsArray.length; i++) {
+            for (int j = 0; j < ballsArray.length; j++) {
```

Comment:
should probably uses bubblesArray.length here. Since they are the same length in this case it doesn't matter, but logically.